### PR TITLE
[BX-8] New Feature: CoPilot first reviewer prompt

### LINKED ISSUE #8: Feature request: CoPilot First Reviewer
## Summary

Finding some reoccurring themes for prompting CoPilot as a "first reviewer" of local pre-commit, working branch code to, need to provide a baseline `prompt` to review code and provide feedback...where possible, including suggested changes which can be applied with a couple of clicks.

## Describe the solution you'd like

- agent prompt(s)

List of inclusion

- evaluate my code for correctness
- evaluate my function docs and type hints for gaps and readability
- evaluate my approach for this functionality.  Suggest alternative approaches.  Look for more efficient ways to achieve the result.
- evaluate my code for quality and maintainability
- analyze this code for refactoring opportunities, look for ways to reduce code or simplify logic
- identify gaps in my unit test scenarios for this code
- evaluate if the log statements are accurate in describing the state of the logic throughout

consider
- user elicitation
- https://docs.github.com/en/copilot/tutorials/customization-library/prompt-files/review-code
- https://github.com/github/awesome-copilot/blob/main/instructions/code-review-generic.instructions.md

## Alternatives you've considered

Copilot code review is a premium feature, available with the Copilot Pro, Copilot Pro+, Copilot Business, and Copilot Enterprise plans. 

https://docs.github.com/en/copilot/concepts/agents/code-review

## Additional context

Should be effective, but succinct, while allowing leads in projects to extend or add specific project level guidance as needed

### PR DESCRIPTION
**Pull request by scottqueen-bixal**
## Description

Tell us what you changed and why.

## Related Issues

Closes #ISSUE_NO

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Other (please explain):

## How did you test?

leverage `test-code` for testing against the `code-review.prompt.md`
>[!NOTE]
>This code will be removed before merged

By default `/code-review` will review uncommitted changes
- [ ] make changes to file, keep unstaged
- [ ] use `/code-review`

Other methods
By selected lines:
- [ ] highlight lines of code in current open file
- [ ] use `/code-review`

By file in context window:
- [ ] drag test files to the chat for context
- [ ] use `/code-review`

## Checklist

- [ ] I've done a self-review of my code
- [ ] I've updated the docs

### CODE REVIEW

File: .github/prompts/code-review.prompt.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+- Can nested loops/conditionals be flattened or extracted?
+- Is there dead code or unused imports/variables?
+- Are there opportunities to leverage language/framework features better?
+
+### 7. Testing & Coverage
+- Are there gaps in unit test scenarios for this code?
+- Are edge cases and error conditions tested?
+- Are tests readable and maintainable?
+- Is test coverage adequate for the risk level?
+- Are integration points properly tested?
```
Comment (LZ6Z-bixal) [inline .github/prompts/code-review.prompt.md]: The testing requirements should be flexible and maybe reference a repo-level document describing testing approach.  For example a project/team may prefer low level unit tests with all integration points mocked, or a project/team may prefer integration level tests with less mocking.  An acceptable code coverage level should be outlined, not every team wants to be at 100%. (The last 5-10% coverage is often a lot of work for questionable value)

File: .github/prompts/code-review.prompt.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+- Is the log level appropriate (debug, info, warn, error)?
+- Are sensitive data (passwords, tokens, PII) excluded from logs?
+- Are important state changes and errors logged?
+
+### 11. Standards & Conventions
+- Does code follow established team/project conventions?
+- Are style guidelines followed (or enforced by linters)?
+- Are there linting/type-checking errors that need addressing?
+- For migrations: Is there a deployment plan documented?
+- Is superfluous or experimental code being committed accidentally?
```
Comment (LZ6Z-bixal) [inline .github/prompts/code-review.prompt.md]: This is mostly duplicate of section 3

File: .github/prompts/code-review.prompt.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+agent: ask
+description: 'Perform comprehensive code review as a senior technical lead, evaluating correctness, quality, maintainability, and adherence to standards with actionable feedback'
+---
+
+```prompt
+# Task: Code Review as Senior Technical Lead
+
+You are a senior technical software engineer lead performing a thorough pre-commit code review. Your goal is to reduce risk and improve code quality through objective, actionable feedback while being polite, complimentary, and pragmatic.
+
+## Review Scope
```
Comment (LZ6Z-bixal) [inline .github/prompts/code-review.prompt.md]: Does the review happen one time, on every commit, on demand?

File: .github/prompts/code-review.prompt.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+Provide your review in this format:
+
+### Summary
+[Brief 2-3 sentence overview of the changes and overall assessment]
+
+### Strengths
+[Highlight 2-3 positive aspects - be genuinely complimentary]
+
+### Required Changes [h: high]
```
Comment (LZ6Z-bixal) [inline .github/prompts/code-review.prompt.md]: Consider if the reviewer should have a limit on the number of comments.  We don't want to have a review seem harsh because of the volume of changes.

Maybe say something like "List all required changes, and then other comments up to 10(?) ordered by priority". The amount is just an example magic number.

File: .github/prompts/code-review.prompt.md (hunk 1 above)
Comment (scottqueen-bixal) [reply .github/prompts/code-review.prompt.md]: I think this is a case where a lead could "tweak" for their own references, but added a third item in `## Review Scope` to generically address.

File: .github/prompts/code-review.prompt.md (hunk 4 above)
Comment (scottqueen-bixal) [reply .github/prompts/code-review.prompt.md]: updated

File: .github/prompts/code-review.prompt.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+- Remember: "Perfect is the enemy of good" - pragmatism over perfection
+- Focus on reducing risk, not producing perfect code
+
+**Reviewer Conduct**:
+- **Address the code, not the developer**: Comments should be about the code itself, never evaluative of the person
+- **Write clear and specific comments**: Be precise about what needs to change and why
+- **Include positive feedback**: Always highlight strengths and good decisions, not just issues
+- **Be generous with code examples**: Show the suggested improvement in code when possible
+- **Frame feedback as requests, not commands**: "Consider refactoring..." vs. "Refactor this..."
+- **Tie notes to principles, not opinions**: Reference standards, best practices, or project guidelines rather than personal preference
```
Comment (mejiaj) [inline .github/prompts/code-review.prompt.md]:
```suggestion
- **Tie notes to principles, not opinions**: Reference standards, best practices, official guidance, or project guidelines rather than personal preference
```

**Suggestion**: Not sure if it needs to be said explicitly, but something like prioritize the official source first.

File: .github/prompts/code-review.prompt.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+### 2. Correctness & Functionality
+- Does the code accomplish its intended purpose?
+- Are there any logical errors or edge cases not handled?
+- Are error conditions properly handled?
+- Does the code match any requirements or specifications?
+
+### 3. Code Quality & Maintainability
+- **System design**: How well do the various pieces interact together? Is the integration with the overall system well-designed?
+- **Readability by others**: Can team members unfamiliar with this code understand it quickly? Is it written for humans first?
```
Comment (mejiaj) [inline .github/prompts/code-review.prompt.md]:
```suggestion
- **Readability for others**: Can team members unfamiliar with this code understand it quickly? Is it written for humans first?
```

**Review by mejiaj — COMMENTED** [review]
Thanks for adding this, I'm excited to try it out! Added my comments from a brief review/test.

File: .github/prompts/code-review.prompt.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+Before completing the review, ask the author:
+
+1. **Change rationale**: "Can you explain the primary reason for these changes? What problem are we solving?"
+
+2. **Alternative approaches**: "Were there other approaches you considered but decided against? What led you to this solution?"
+   - This provides context and prevents reviewers from suggesting already-explored alternatives
+
+3. **Risk assessment**: "Are there any areas of this code you're uncertain about or would like specific feedback on?"
+
+4. **Deployment considerations**: "Are there any special deployment steps, feature flags, or rollback plans needed for these changes?"
```
Comment (mejiaj) [inline .github/prompts/code-review.prompt.md]: Simplifying because there were a lot of questions when I was reviewing multiple features.

```suggestion
### Questions for Author

Only include questions not already answered by the diff. Consider:

- What problem does this solve, and why this approach?
- Were other approaches considered?
- Are there areas you'd like specific feedback on?
- Are there deployment steps, feature flags, or rollback plans needed?
```

**Review by mejiaj — COMMENTED** [review]
@scottqueen-bixal added some suggestions in areas where I condensed after testing in different projects.

**Review by mejiaj — APPROVED** [review]

**Comment by scottqueen-bixal** [comment]
> @scottqueen-bixal added some suggestions in areas where I condensed after testing in different projects.

@mejiaj - Thanks for proposing these recommended prompt compressions. I've committed those in, tested, and removed our testing assets.

This is good to merge.
